nexus/saga-recovery/src/recovery.rs

```diff
@@ -0,0 +1,705 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Guts of the saga recovery bookkeeping
+
+use super::status::RecoveryFailure;
+use super::status::RecoverySuccess;
+use chrono::{DateTime, Utc};
+use omicron_common::api::external::Error;
+use slog::{debug, error, info, warn};
+use slog_error_chain::InlineErrorChain;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use steno::SagaId;
+use tokio::sync::mpsc;
+
+/// Describes state related to saga recovery that needs to be maintained across
+/// multiple passes
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct RestState {
+    /// set of sagas that we believe may be running
+    ///
+    /// See the big block comment in the saga recovery background task for more
+    /// on how this works and why.
+    sagas_started: BTreeMap<SagaId, SagaStartInfo>,
+    remove_next: BTreeSet<SagaId>,
+}
+
+/// Describes how we learned that a particular saga might be running
+///
+/// This is only intended for debugging.
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[allow(dead_code)]
+struct SagaStartInfo {
+    time_observed: DateTime<Utc>,
+    source: SagaStartSource,
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+enum SagaStartSource {
+    StartChannel,
+    Recovered,
+}
+
+impl RestState {
+    /// Returns initial at-rest state related to saga recovery
+    pub fn new() -> RestState {
+        RestState {
+            sagas_started: BTreeMap::new(),
+            remove_next: BTreeSet::new(),
+        }
+    }
+
+    /// Read messages from the channel (signaling sagas that have started
+    /// running) and update the set of sagas that we believe may be running.
+    ///
+    /// See the big block comment in the saga recovery background task for more
+    /// on how this works and why.
+    pub fn update_started_sagas(
+        &mut self,
+        log: &slog::Logger,
+        sagas_started_rx: &mut mpsc::UnboundedReceiver<SagaId>,
+    ) -> usize {
+        let (new_sagas, disconnected) = read_all_from_channel(sagas_started_rx);
+        if disconnected {
+            warn!(
+                log,
+                "sagas_started_rx disconnected (is Nexus shutting down?)"
+            );
+        }
+
+        let rv = new_sagas.len();
```

Comment:
was going to suggest calling `self.sagas_started.reserve(rv)` here to reallocate only once, but then i remembered `BTreeMap` doesn't have that, because it's a tree...only `HashMap` can do that. never mind!